docs/design/sql/20-matching-sql-query-to-jet-job.md

```diff
@@ -0,0 +1,237 @@
+was chosen as main direction to implement.
+
+### Solution design
+
+#### Existing client - member communication protocol
+
+![Client - Member communication protocol](https://svgshare.com/i/z9v.svg)
+
+#### Client protocol
+
+To enable asynchronous query handling, we propose adding a field to the `SqlExecuteCodec.RequestParameters` to mark
+requests that should be handled asynchronously. In order to support [Flexible Async Mode](#flexible-async-mode)
+it will be an integer field with the following values:
+
+- `-1` - synchronous mode (default and for backwards compatibility)
+- `0` - asynchronous mode
+- `> 0` - timeout for flexible async mode before returning async response
+
+Initially only the first 2 values will be supported. Flexible async mode may be implemented in the future if
+the performance tests confirm that latency for short DML queries might be improved that way.
+
+#### Java APIs
+
+The proposed Java API changes are as follows:
+
+```java
+public interface SqlService {
+// ...
+
+    /**
+     * Execute the statement in asynchronous way if possible.
+     * @param statement
+     * @return
+     */
+    // also additional convenience variants similar to `execute`
+    // supported only on client-side (SqlClientService).
+    @Nonnull
+    SqlResult executeAsync(@Nonnull SqlStatement statement);
+}
+
+public interface AsyncSqlResult extends SqlResult {
+    /**
+     * @return If the query uses/used Jet job
+     */
+    boolean isJob();
+
+    /**
+     * Returns job id of job used to execute this query.
+     * <p>
+     * If the job was a light job (interactive query without {@code ANALYZE})
+     * it will be impossible to find the job after the query finished
+     * which requires caution or error handling to avoid race conditions.
+     *
+     * @throws UnsupportedOperationException if the query did not a use Jet job (isJob() returns false)
+     */
+    long getJobId();
+
+    /**
+     * Wait for query finish.
+     * If the job does not exist it assumes that it already finished.
+     *
+     * @return Final result of the query. It might be `this` or different object containing eg. final `updateCount` 
+     * @throws UnsupportedOperationException if the query did not a use Jet job (isJob() returns false)
+     */
+    SqlResult join();
+}
+
+// Only client-side API implements async execution mode
+public class SqlClientResult implements AsyncSqlResult {
+// ...
+}
+```
+
+Within the SqlService, we're implementing a new method named `executeAsync`. When using this method, the client can
+indicate that a query should be executed asynchronously by setting an asynchronous flag. Afterward, `execute` task
+is sent via calling `executeAsync`, the client will receive an initial `AsyncSqlResult` as part of the
+`SqlExecuteResponse`. This result will include the `jobId` of the corresponding Jet job. If the query produces rows,
```

Comment:
`jobId` and `AsyncSqlResult` is needed both for SELECT and DML. It is worth mentioning explicitly